src/mongo/jobs/DiscoverOutdatedComposites.class.php

```diff
@@ -0,0 +1,296 @@
+    }
+
+    
+    public function runRegenerationTasks($regenTasks) {
+        foreach ($regenTasks as $regenTask) {
+            $specification = $regenTask->specification;
+            $compositeCollection = $regenTask->compositeCollection;
+            $compositeRegenFunction = $regenTask->compositeRegenFunction;
+            
+            foreach($regenTask->cbdDocuments as $cbdDoc) {
+                $compositeRegenFunction($specification, $compositeCollection, $cbdDoc);
+            }
+        }
+    }
+
+    /**
+     * Generates regeneration task data for any outdated documents in a composite collection,
+     * by scanning that composite collection one cursor-page at a time.
+     *
+     * @param CompositeMetadata $metadatum - the composite collection to search
+     * @return CompositeRegenTask|null - a CompositeRegenTask, if outdated documents are found; null otherwise
+    **/
+    public function getRegenTaskForMetadata($metadatum, $cursorLimit) {
+        $filterOutdated = $metadatum->getOutdatedQueryComponent();
+        if(!isset($filterOutdated)) {
+            return null;
+        } else {
+            $compositeCollection = $metadatum->compositeCollection;
+            $specification = $metadatum->specification;
+            $cbdCollection = $metadatum->cbdCollection;
+            $compositeRegenFunction = $metadatum->compositeRegenFunction;
+
+            $cbdIdFields = array(_ID_KEY.'.'._ID_RESOURCE => 1, _ID_KEY.'.'._ID_CONTEXT => 1);
+
+            // find one cursor page at a time for this composite
+            $outdatedComposites = 
+                $compositeCollection
+                    ->find($filterOutdated, $cbdIdFields)
+                    ->limit($cursorLimit);
+
+            if(!$outdatedComposites->hasNext()) {
+                // no outdated composite documents found, so no work to be done
+                return null;
+            } else {
+                // find IDs root CBDs required to regenerate the composite
+                $outdatedCbdIds = 
+                    array_map(array($this, 'composite2cbdId'), iterator_to_array($outdatedComposites, false));
+
+                // fetch the CBDs themselves
+                $filterCbdsById = array('$or' => $outdatedCbdIds);
+                $cbdDocs = $cbdCollection->find($filterCbdsById);
+
+                if(!$cbdDocs->hasNext()) {
+                    // no CBDs found, so nothing to regenerate
+                    return null;
+                } else {
+                    // return a task to regenerate outdated composite documents
+                    return new CompositeRegenTask(
+                        $compositeRegenFunction, 
+                        $specification, 
+                        $compositeCollection, 
+                        iterator_to_array($cbdDocs, false)
+                    );
+                }
+            }
+        }
+    }
+
+    // TODO: this should really live in IComposite implementations, but cannot yet.
```

Comment:
This entire function should really live in an `IComposite` implementation, like `Views.class.php`.  However, I've put it here for now so we can discuss my concerns.

It's currently difficult to move it there because their constructors are tied to collections that seem independent of all views, and contexts that seem like a function-level specification.